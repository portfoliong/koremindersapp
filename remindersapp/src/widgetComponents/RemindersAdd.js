import React, { Component } from 'react';
import imageOne from '../images/addRemindersChecklistImage.jpg';
class RemindersAdd extends Component{
    state = {
        id: null,
        message: null,
        

    }
    handleInput = (e) => {
        //Set state functiom takes 2 parameters
            // What property that needs changing, what value to chand it with
            // Remember, target = the object that fired the event 
            // So, 
                // event, in the input box: that has the id 'NAME' = e.target.id
                // event, in the inout box: what is the value in the box = e.targer.value
            // this.setState({[NAME]:ValueInNameBox)
          
            this.setState({[e.target.id]: e.target.value})
        
    }

    handleSubmit = (e) => {
        //To prevent the page reloading, preventDefault
        e.preventDefault();
        // Below is very important ...
        // Remeber, we've passed the 'addReminderToState' function as a prop
        // Thus allowing us to call the method from this class (addReminders)
        // The method (addReminderToState), asks for a single prop
        // We pass in the 'state', which is the details of the 'champ' that the user ...
        // sumbited in the form
        // Alas, the addReminderToState method IN THE APP.JS gets passed our champ.
        this.props.addReminderToState(this.state)
        // this.props.addChampsToState(this.state)

        //Removing input from textbox after it's been added
        this.setState({message:''})
    }
    render(){
        return(
                        
                <div className='card'>
                    <div className="card-image">
                        <img src={imageOne}/>
                        <span className='finishedRemindersList card-title'>Add Reminders</span>
                    </div>

                                
                        <div className='card-content'>
                            <form className="waves-effect waves-light" onSubmit={this.handleSubmit }>
                                <label htmlFor="name">Message</label>
                                {/* 'required' tag put in to ensure the reminder isn't left blank */}
                                <input  required placeholder="Hey! Add a Reminder" value={this.state.message} type="text" id="message" onChange={this.handleInput}/>

                                <button className="waves-effect waves-light btn z-depth-1 pulse #b71c1c red darken-4">Submit</button>

                                <div><hr/></div>
                            </form>
                        </div>
                </div>
        )
    }
}
export default RemindersAdd



