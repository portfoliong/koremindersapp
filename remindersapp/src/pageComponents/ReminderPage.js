import React, { Component } from 'react';
import Reminders from '../widgetComponents/Reminders.js';
import RemindersAdd from '../widgetComponents/RemindersAdd.js';
import FinishedReminders from '../widgetComponents/FinishedReminders.js';

class ReminderPage extends Component{
  state = {
      reminders: [
        {id: 1, message:'Do some more React'},
        {id: 2, message:'Or Javascript'},
        {id: 3, message:'But probably just react'}
      ],
      
      finishedReminders: [
        {id: null , message:null},
      ]
}
  


  // PROPS can be used for more than just data.
  // We can pass the below FUNCTION as a prop.
  // This allows the component that was passed the below function ...
// To have access to it.
addReminderToState = (newReminder) => {
    // We're just gonna add a random ID (uniqueKey) to our champ object so it  can go in our state above
    newReminder.id = Math.random();
    
    //We can't change  the state directly, that's desctructive. 
    //So, instead we're going to use  the spread operator, to create a new array
    //The spread operator takes each element in the array, and allows you to append another element (i.e. our newChamp)
      // Syntax: let newArray = [...this.state.oldArray, newItem]
      // Note: the '...' and the square brackets ([]) are syntax for the SPREAD operator
    let reminders = [...this.state.reminders, newReminder];

    //Then, we equate our new array (champs ... i CHOSE the same name), to  our  state
      //by using the set state  function
      // notice that we don't write  champs = champs, we write champs: champs
    this.setState({
      reminders: reminders
    })



}

removeReminderFromState = (id) => {
  //As we know from the above methoer (addReminderToState), we cannot have any destructive editing of the state
  //So, we have to create a new aray, and setState with that new array. 
  //It reads, create a new array called reminders. take out any reminders in reminders that has a uniqueKey the same as the one passed in via the parameter
  let reminders = this.state.reminders.filter(reminders => {
     return reminders.id !== id
   })

   this.setState({reminders: reminders})

   console.log(reminders)
}

addFinishedReminder = (reminders) =>{
// Follows same explanation as 'addReminderToState'.
// Creating an array that appends the new 'finished reminder'
   let completedReminders = [...this.state.finishedReminders, reminders];

     this.setState({
       finishedReminders: completedReminders
     })

     let thisNewFinishedRemindersMessage = reminders.message;

//Ensuring the finished reminder is removed from reminder.
// Why?: If the finished reminder is removed from the 'active' reminders list,  it cannot be added twice by the  user 
    let updatedReminders = this.state.reminders.filter(reminders => {
      return reminders.message !== thisNewFinishedRemindersMessage
    })

    this.setState({reminders: updatedReminders})
  console.log(this.state.reminders)
  
  
    
}

removeFinishedReminder = () =>{
//Follows same explanation as 'removeReminderFromState'.
 let completedReminder = '';
 this.setState({finishedReminders: ''})
}

 
  render(){
    return (
            <div className="App row center">
                <div className="col s12 m4 l2"><h5>Add a Reminder</h5>
                    <RemindersAdd  addReminderToState={this.addReminderToState}/>
                </div>

                <div className="col s12 m4 l8"><h5>Reminders</h5>
                    <Reminders  reminders={this.state.reminders} removeReminder={this.removeReminderFromState}  addFinishedReminder={this.addFinishedReminder}/>
                </div>
                
                <div className="col s12 m4 l2"><h5>Complete Reminders</h5>
                    <FinishedReminders  removeFinishedReminder={this.removeFinishedReminder} finishedReminders={this.state.finishedReminders}/>
                </div>
            </div>
              
    )    
    
    }
}
export default ReminderPage;
