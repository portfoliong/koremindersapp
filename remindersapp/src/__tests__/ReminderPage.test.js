// ReminderPage.test.js
import React from 'react';
import { mount, shallow } from 'enzyme';
import ReminderPage from '../pageComponents/ReminderPage.js';
import RemindersAdd from '../widgetComponents/RemindersAdd';
import renderer from 'react-test-renderer';

// describe is used as a 'container' for the tests
//     --> Think of it as a functional component, that exuctes all tests inside it
//     --> It takes 2 parameters, a name, and a function
//     --> Note the unkown function synax (name, unknown function {function contents})
describe ('Test: Initialise Reminder page and mount all React components', () =>{

    // --> Again, note the syntax. name/unknown function/ function contents
    //below is a Unit test (: a test for ONE specific function)
    it('loads the reminder page',() =>{
        shallow(<ReminderPage />);
    });

    //below is an Integration test (: a test for ALL components in the tree of the reminder page (related))
    it('loads all components needed for the reminder page',() => {
        mount(<ReminderPage />);
    });

    //snapshot: 
    // --> this imported utility allows for the testing of the output of the render method. i.e. is what I want rendered, rendered accordindly.
    // --> It reads: expect this object (reminderPage component) to match the snapshot
    it('tests to see if the render method in ReminderPage renders correctly', () => {
        const reminderPage = shallow(<ReminderPage/>);
        expect(reminderPage).toMatchSnapshot();
    })

    //event tests:
    // --> below Eznymes mount and simulate methods are going to be use to: load the reminder page, then see if the submission (Add reminder) button works
    it('test the add ReminderAdd sumbit button works',()=>{
        const reminderAdd = mount(<RemindersAdd/>);
        //we create a variable of the component. Then 'in that page', we find an element 'button', with the id of '#remindersAddSubmitButton'
        //we then simulate a 'click' on the button
        reminderAdd
            .find('button#remindersAddSubmitButton')
            .simulate('click')

        reminderAdd.unmount();
    })
    
});

//It also worth noting that Jest is a node-base TEST RUNNER, while Enzyme is a TEST UTILITY PACKAGE.
//  --> i.e. the methods 'shallow' & 'mount' are Enzyme, the rest (it, describe) are Jest. 