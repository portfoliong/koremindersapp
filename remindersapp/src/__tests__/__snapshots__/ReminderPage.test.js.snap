// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test: Initialise Reminder page and mount all React components tests to see if the render method in ReminderPage renders correctly 1`] = `ShallowWrapper {}`;
